# Requirement R1

In [ ]:
import pandas as pd
url = "https://raw.githubusercontent.com/jbrownlee/Datasets/master/pima-indians-diabetes.data.csv"
names = ['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class']
rawdata = pd.read_csv(url, names=names)

In [ ]:
rawdata.columns

Index(['preg', 'plas', 'pres', 'skin', 'test', 'mass', 'pedi', 'age', 'class'], dtype='object')

In [ ]:
print(rawdata)

     preg  plas  pres  skin  test  mass   pedi  age  class
0       6   148    72    35     0  33.6  0.627   50      1
1       1    85    66    29     0  26.6  0.351   31      0
2       8   183    64     0     0  23.3  0.672   32      1
3       1    89    66    23    94  28.1  0.167   21      0
4       0   137    40    35   168  43.1  2.288   33      1
..    ...   ...   ...   ...   ...   ...    ...  ...    ...
763    10   101    76    48   180  32.9  0.171   63      0
764     2   122    70    27     0  36.8  0.340   27      0
765     5   121    72    23   112  26.2  0.245   30      0
766     1   126    60     0     0  30.1  0.349   47      1
767     1    93    70    31     0  30.4  0.315   23      0

[768 rows x 9 columns]


In [ ]:
rawdata.isnull().sum()

preg     0
plas     0
pres     0
skin     0
test     0
mass     0
pedi     0
age      0
class    0
dtype: int64

In [ ]:
x = rawdata.iloc[:, :-1].values
y = rawdata.iloc[:, -1].values

In [ ]:
# minmaxscaled x
from sklearn.preprocessing import MinMaxScaler
mm = MinMaxScaler()
x = mm.fit_transform(x)


**Splitting data into 60% for training , 20% for testing , 20% for validation**



In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_temp, y_train, y_temp = train_test_split(x, y, test_size=0.4, random_state=20)
x_test, x_val, y_test, y_val = train_test_split(x_temp, y_temp, test_size=0.5, random_state=20)


In [ ]:
from collections import Counter
Counter(y_train)

Counter({0: 298, 1: 162})

In [ ]:
print(x_train)

[[0.35294118 0.46231156 0.75409836 ... 0.29657228 0.0469684  0.11666667]
 [0.23529412 0.47236181 0.53278689 ... 0.3681073  0.02988898 0.        ]
 [0.23529412 0.41708543 0.70491803 ... 0.4366617  0.10204953 0.21666667]
 ...
 [0.11764706 0.54271357 0.50819672 ... 0.37555887 0.02134927 0.        ]
 [0.23529412 0.57286432 0.52459016 ... 0.43070045 0.0204953  0.05      ]
 [0.52941176 0.82914573 0.72131148 ... 0.45305514 0.09564475 0.46666667]]


# Requirement R2


In [ ]:
from sklearn.linear_model import Perceptron
ppt1 = Perceptron()
ppt1.fit(x_train,y_train)


Perceptron()

In [ ]:
y_pred = ppt1.predict(x_test)

In [ ]:
from sklearn.metrics import f1_score
f1_class_0 = f1_score(y_test, y_pred, pos_label=0)
f1_class_1 = f1_score(y_test, y_pred, pos_label=1)

# Printed F1 scores for each class
print(f" Class 0: {f1_class_0}")
print(f" Class 1: {f1_class_1}")


 Class 0: 0.8244897959183672
 Class 1: 0.31746031746031744


# Requirement R3

In [ ]:
from imblearn.over_sampling import SMOTE
smm = SMOTE(sampling_strategy=1.0, k_neighbors=5)
x_smm,y_smm = smm.fit_resample(x_train,y_train)

In [ ]:
from collections import Counter
Counter(y_smm)

Counter({0: 298, 1: 298})

In [ ]:
from sklearn.linear_model import Perceptron
ppt2 = Perceptron()

In [ ]:
ppt2.fit(x_smm,y_smm)

Perceptron()

In [ ]:
y_pred_smm = ppt2.predict(x_test)

In [ ]:
from sklearn.metrics import f1_score
f1_class_0 = f1_score(y_test, y_pred_smm, pos_label=0)
f1_class_1 = f1_score(y_test, y_pred_smm, pos_label=1)

# Printed F1 scores for each class
print(f" Class 0: {f1_class_0}")
print(f" Class 1: {f1_class_1}")

 Class 0: 0.8279069767441861
 Class 1: 0.6021505376344087


# Requirement R4

Before SMOTE:
Class 0: 0.8244
Class 1: 0.3174
After SMOTE:
Class 0: 0.8279
Class 1: 0.6021

The increase in the F-measure for Class 1 indicates that the model’s performance on the minority class has improved

# Requirement R5

In [ ]:
ppt3 = Perceptron()
ppt3.fit(x_smm,y_smm)
importance = ppt3.coef_
print(importance)


[[ 2.27559904  6.34469125 -1.30341786  0.34869594 -0.41494102  6.90001992
   2.30621115 -0.86717363]]


By examining these coefficients, we can gauge which features have a weak influence on the perceptron's decision-making process. Features with least coefficient magnitudes typically have a least impact on the model's predictions.
two least significant predictors are index 2,4 ['pres','test']





In [ ]:
# two least significant predictors
columns_to_drop = rawdata.columns[[2, 4]]
new_df = rawdata.drop(columns_to_drop, axis=1)


In [ ]:
new_df

preg  plas  skin  mass   pedi  age  class
0       6   148    35  33.6  0.627   50      1
1       1    85    29  26.6  0.351   31      0
2       8   183     0  23.3  0.672   32      1
3       1    89    23  28.1  0.167   21      0
4       0   137    35  43.1  2.288   33      1
..    ...   ...   ...   ...    ...  ...    ...
763    10   101    48  32.9  0.171   63      0
764     2   122    27  36.8  0.340   27      0
765     5   121    23  26.2  0.245   30      0
766     1   126     0  30.1  0.349   47      1
767     1    93    31  30.4  0.315   23      0

[768 rows x 7 columns]

In [ ]:
a = new_df.iloc[:, :-1].values
b = new_df.iloc[:, -1].values

In [ ]:
# minmaxscaled a
from sklearn.preprocessing import MinMaxScaler
mm2 = MinMaxScaler()
a = mm2.fit_transform(a)


In [ ]:
from sklearn.model_selection import train_test_split

a_train, a_temp, b_train, b_temp = train_test_split(a, b, test_size=0.4, random_state=0)
a_test, a_val, b_test, b_val = train_test_split(a_temp, b_temp, test_size=0.5, random_state=0)

In [ ]:
from collections import Counter
Counter(b_train)

Counter({0: 295, 1: 165})

In [ ]:
print(a_train)

[[0.17647059 0.65326633 0.23232323 0.42324888 0.10461144 0.21666667]
 [0.05882353 0.40201005 0.11111111 0.44709389 0.19171648 0.01666667]
 [0.23529412 0.72361809 0.28282828 0.43964232 0.08923997 0.26666667]
 ...
 [0.23529412 0.47236181 0.22222222 0.3681073  0.02988898 0.        ]
 [0.64705882 0.42713568 0.         0.4485842  0.09479078 0.23333333]
 [0.29411765 0.68341709 0.         0.         0.23996584 0.8       ]]


In [ ]:
from sklearn.linear_model import Perceptron
ppt4 = Perceptron()
ppt4.fit(a_train,b_train)

Perceptron()

In [ ]:
y_pred_new_df = ppt4.predict(a_test)

In [ ]:
from sklearn.metrics import f1_score
f1_class_0 = f1_score(b_test, y_pred_new_df, pos_label=0)
f1_class_1 = f1_score(b_test, y_pred_new_df, pos_label=1)

# Printed F1 scores for each class
print(f" Class 0: {f1_class_0}")
print(f" Class 1: {f1_class_1}")

 Class 0: 0.7817258883248732
 Class 1: 0.6126126126126126


# Requirement R6

In [ ]:
ppt5 = Perceptron()
ppt5.fit(a_train,b_train)
importance = ppt5.coef_
print(importance)

[[2.11764706 3.39698492 0.53535354 3.36363636 2.10674637 0.88333333]]


By examining these coefficients, we can gauge which features have a stronger influence on the perceptron's decision-making process. Features with larger coefficient magnitudes typically have a greater impact on the model's predictions. The two most significant predictors are plas with index 1 and mass with index 3

# Requirement R7

In [ ]:
from sklearn.model_selection import GridSearchCV
from imblearn.pipeline import Pipeline
from imblearn.over_sampling import SMOTE
from sklearn.linear_model import Perceptron

# uesd pipeline
model = Perceptron()
resampling = SMOTE()
pipeline = Pipeline([('SMOTE', resampling), ('Perceptron', model)])

#parameter grid
param_grid = {
    'SMOTE__sampling_strategy': [0.5, 0.75, 1.0],
    'SMOTE__k_neighbors': [3, 5, 7]
}


grid_search = GridSearchCV(estimator=pipeline, param_grid=param_grid, cv=5, scoring='f1')


grid_search.fit(a_val, b_val)


best_params = grid_search.best_params_
print(f"Best parameters: {best_params}")


Best parameters: {'SMOTE__k_neighbors': 5, 'SMOTE__sampling_strategy': 1.0}


In [ ]:
# Generate new perceptron model with optimal Solution
smote_optimal = SMOTE(sampling_strategy=best_params['SMOTE__sampling_strategy'], k_neighbors=best_params['SMOTE__k_neighbors'])
a_train_res, b_train_res = smote_optimal.fit_resample(a_train, b_train)

ppt_optimal = Perceptron()
ppt_optimal.fit(a_train_res, b_train_res)


Perceptron()

In [ ]:
y_pred_optimal = ppt_optimal.predict(a_test)

f1_class_0_optimal = f1_score(b_test, y_pred_optimal, pos_label=0)
f1_class_1_optimal = f1_score(b_test, y_pred_optimal, pos_label=1)

print(f"Optimal F1 scores:")
print(f" Class 0: {f1_class_0_optimal}")
print(f" Class 1: {f1_class_1_optimal}")


Optimal F1 scores:
 Class 0: 0.7176470588235295
 Class 1: 0.6521739130434783
